dvc/repo/brancher.py

```diff
@@ -77,6 +77,7 @@ def brancher(  # noqa: E302
         from dvc.fs import GitFileSystem
 
         for sha, names in found_revs.items():
+            self.__dict__.pop("index", None)
```

Comment:
This at least makes `repo.index` usable at the same time as brancher. Previously you would just reuse whatever the first built index was due to it being a cached_property (so you would usually get the workspace index no matter what brancher rev you were at)

Reply:
I think what we really want is a brancher replacement that returns built indexes, and ideally caches them on disk by git SHA. So for known git SHA's that we have already indexed we can just load the cached version instead of re-collecting the entire DVC repo for each rev every time we use brancher.

Reply:
> Previously you would just reuse whatever the first built index was due to it being a cached_property

Not sure I understand this. Index is cleared when the `fs` changes.

https://github.com/iterative/dvc/blob/56066088678b45315e6a8f88c4fd42294e36c977/dvc/repo/__init__.py#L321-L325
https://github.com/iterative/dvc/blob/56066088678b45315e6a8f88c4fd42294e36c977/dvc/repo/__init__.py#L576-L580

> I think what we really want is a brancher replacement that returns built indexes

That was the original idea, see this [gist](https://gist.github.com/skshetry/052a69b8dbeea5564d4c65fb73272e30) for example. The problem is the way we are swapping out `fs` in between which needs to be fixed and needs to be per-index.

> ideally caches them on disk by git SHA

That was something that I was thinking of working on next (after introducing index). So we have these bits implemented:

https://github.com/iterative/dvc/blob/56066088678b45315e6a8f88c4fd42294e36c977/dvc/repo/index.py#L307
https://github.com/iterative/dvc/blob/56066088678b45315e6a8f88c4fd42294e36c977/dvc/repo/index.py#L319

And the reason why `index` is immutable. But there were lots of issues, the important one being that the data structure was not appropriate at that time and the thinking in terms of `stage` is not quite suitable (if `data` is what we cared about). It is definitely complicated since we have params, metrics, etc which could also be cached.

Since the implementation currently uses `stage.dumpd()`, it is terribly slow to even create automatically. Last time, it would take ~5s just to serialize for a significantly large repo. Also there's a lot of cache invalidation that we need to be careful of (eg: `dvc.yaml`/`.dvc` broken, metrics missing, etc.)

Reply:
I missed the fs setter/_reset behavior. I was seeing some issue with caching during brancher but it might have been something unrelated then. I'll take another look and then remove the brancher change if I can't reproduce it again

Reply:
I think it can be removed, I'll take care of it when the import-url changes get merged. (I'm still in the process of testing some potential brancher related object+import collection changes)